sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -148,6 +150,12 @@
 abstract class ReadFromKafkaDoFn<K, V>
     extends DoFn<KafkaSourceDescriptor, KV<KafkaSourceDescriptor, KafkaRecord<K, V>>> {
 
+  private static final int OFFSET_UPDATE_INTERVAL_SECONDS = 1;
+
+  private transient ScheduledExecutorService backlogFetcherThread =
```

Comment:
this is per-dofn, it seems like it should be some static cache across all of the dofns.  

You could see what I was doing for the caching the background reading threads in https://github.com/apache/beam/pull/31786/files#diff-0f4b915b82782e618addba2e443f07b3d4b86ad8b77c2bf7adeb0ad1d6864adb

Reply:
Note that this is a little different than the UnboundedReader where there is a single reader instance for each partition+topic and the caching across dofns is managed by the runner with the ReaderCache.  Since these are just normal dofns, there may be many different DoFn instances (up to 1-per processing thread) and the same partition+topic may bounce between using different dofn instances.